### DISCUSSION TITLE
Welcome to MonoAlg3D_C Discussions!

### DISCUSSION THREAD

**Discussion by rsachetto** [rsachetto/MonoAlg3D_C#30] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Aditi1993** [comment]
Hi 
How can I achieve the following?
Lets say, we are using example_configs/cable_mesh_with_CRN_atrial.ini and i want to change the variable "period" and run multiple simulations and output them in different "output_dir". But i do not want to generate multiple  cable_mesh_with_CRN_atrial.ini  files.

I found a way to do this with replacing lines with a bash script (below). Please let me know if there is another way for this.
Thank you
 
#!/bin/bash

array=( 0 200.0 400.0 600.0 800.0 )
array2=( 0 2000.0 4000.0 6000.0 8000.0 )

for i in {1..4}; do
    #echo "simulation_time=${array2[i-1]}"
    sed -i -e "s/simulation_time=${array2[i-1]}/simulation_time=${array2[i]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
    sed -i -e "s/cableCRN_${array[i-1]}/cableCRN_${array[i]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
    sed -i -e "s/period=${array[i-1]}/period=${array[i]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
    sbatch run_benchmark.sh
done

**Reply by rsachetto** [reply]
Hi,
It is possible to change the output_dir by passing the following command-line switch:

./bin/MonoAlg3D -c example_configs/cable_mesh_with_CRN_atrial.ini --save_result "output_dir=outputs/test"

Unfortunately, it is not possible to change stimuli sections options using the command line yet. I will try to implement this feature and I will also improve the simulator help, as some of the options do not appear in the help.

As you are using the cluster, I think your solution would be more suitable, because with sbatch you need to change the parameters before each run submission. So, even with the command line switches, you would still have to change the launch script parameters for each simulation. 

Also, check if your solution is correct, as it changes the config file and then submits the job. I think that only the last configuration will be executed as the jobs are not scheduled immediately. I think you would have to create a ini file for each simulation and also change the run_benchmark.sh to use the copied configuration.

**Reply by Aditi1993** [reply]
yes, you are right, i found that out the hard way and had to introduce a delay and generate new config file. Thank you for the output directory option.

#!/bin/bash

array=( 0 200 400 600 800 )
array2=( 0 2000.0 4000.0 6000.0 8000.0 )

sed -i -e "s/simulation_time=800.0/simulation_time=${array2[0]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
sed -i -e "s/cableCRN_800/cableCRN_${array[0]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
sed -i -e "s/period=800/period=${array[0]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
sed -i -e "s/Benchmark_800/Benchmark_${array[0]}/g" run_benchmark.sh
sed -i -e "s/cable_mesh_with_CRN_atrial_800.ini/cable_mesh_with_CRN_atrial_${array[0]}.ini/g" run_benchmark.sh
sleep 10

for i in {1..4};
do
    sed -i -e "s/simulation_time=${array2[i-1]}/simulation_time=${array2[i]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
    sed -i -e "s/cableCRN_${array[i-1]}/cableCRN_${array[i]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
    sed -i -e "s/period=${array[i-1]}/period=${array[i]}/g" example_configs/cable_mesh_with_CRN_atrial.ini
    sed -i -e "s/Benchmark_${array[i-1]}/Benchmark_${array[i]}/g" run_benchmark.sh
    #head -13 example_configs/cable_mesh_with_CRN_atrial.ini    
    cp example_configs/cable_mesh_with_CRN_atrial.ini es/cable_mesh_with_CRN_atrial_${array[i]}.ini
    sed -i -e "s/cable_mesh_with_CRN_atrial_${array[i-1]}.ini/cable_mesh_with_CRN_atrial_${array[i]}.ini/g" run_benchmark.sh
    sleep 10
    sbatch run_benchmark.sh   
done

**Comment by LLRiebel** [comment]
Hi

I have been using the save and restore state functionality in combination with the save_with_activation_times function. During the save state run everything works well with the .acm file looking as expected. However, when using restore state, only the activation times are outputted correctly with the action potential durations all wrongly set to 0. Have you come across this issue at all? I have tried running a second at the beginning of restore state before stimulation in case it was an issue with the apd threshold but got the same result.

Thanks!
Luana

**Reply by rsachetto** [reply]
Hi Luana, 
Thanks for letting me know. Could you please send me a minimal reproducible example, (an ini with a small simulation with the same issue) so I can investigate the problem?

My best,

Rafael Sachetto

**Reply by LLRiebel** [reply]
Thanks for your swift reply! Here's a sample .ini file using a simple box mesh:

[main]
num_threads=12
dt_pde=0.05
simulation_time=1000.0
;simulation_time=3000.0
abort_on_no_activity=false
use_adaptivity=false

[update_monodomain]
main_function=update_monodomain_default

[save_result]
print_rate=100
mesh_print_rate=1000
output_dir=outputs/test
remove_old_simulations=true
apd_threshold=-63
main_function=save_with_activation_times
init_function=init_save_with_activation_times
end_function=end_save_with_activation_times
time_threshold=10.0
file_prefix=V
binary=false

[save_state]
save_rate=19980
main_function=save_simulation_state

;[restore_state]
;main_function=restore_simulation_state

[assembly_matrix]
init_function=set_initial_conditions_fvm
sigma_l=0.0003
sigma_t=0.0001
sigma_n=0.00005
library_file=shared_libs/libdefault_matrix_assembly.so
main_function=anisotropic_sigma_assembly_matrix

[linear_system_solver]
tolerance=1e-15
use_preconditioner=no
max_iterations=100
library_file=shared_libs/libdefault_linear_system_solver.so
use_gpu=yes
main_function=gpu_biconjugate_gradient
init_function=init_gpu_biconjugate_gradient

[domain]
name = Benchmark 05 Mesh
start_discretization = 500.0
maximum_discretization = 500.0
number_of_points = ?
main_function=initialize_grid_with_custom_mesh
mesh_file=meshes/test.alg

[ode_solver]
adaptive=true
dt=0.0001
use_gpu=true
gpu_id=0
library_file= shared_libs/libToRORd_fkatp_endo.so

[stim_plain_0]
start = 0.0
duration = 2.0
current = -50.0
min_x = 0.0
max_x = 1500.0
min_y = 0.0
max_y = 1500.0
min_z = 0.0
max_z = 1500.0
main_function=stim_x_y_z_limits

[stim_plain_1]
start = 2000.0
duration = 2.0
current = -50.0
min_x = 0.0
max_x = 1500.0
min_y = 0.0
max_y = 1500.0
min_z = 0.0
max_z = 1500.0
main_function=stim_x_y_z_limits

**Reply by rsachetto** [reply]
Thanks, I had to create new save and restore state functions as the default ones do not save state from other modules (in this case, we needed the state of the save_mesh module).  I suggest you to make a backup of you current simulator version and pull or clone this new version.

To use these new functions change the save and restore state sections main_function to  save_simulation_state_with_activation_times and restore_simulation_state_with_activation_times. Also, if you now omit the save_rate parameter, the state is only saved in the end of the simulation.

I had to change the behavior of the save_with_activation_times function. Now, if a cell is activated but the simulation finishes before it goes to rest, the APD is not computed, only the activation time. Let me now if you need this information, so I can think about a way to restore these  behavior.

I am sending small test simulation configuration. Could you please test it and tell me if everything makes sense?

[main]
num_threads=6
dt_pde=0.05
simulation_time=650.0
abort_on_no_activity=false
use_adaptivity=false

[update_monodomain]
main_function=update_monodomain_default

[save_result]
print_rate=100
mesh_print_rate=100
output_dir=outputs/test
remove_old_simulations=true
apd_threshold=-63
main_function=save_with_activation_times
init_function=init_save_with_activation_times
end_function=end_save_with_activation_times
time_threshold=10.0
file_prefix=V
binary=false

[save_state]
save_rate=100
main_function=save_simulation_state_with_activation_times

#[restore_state]
#main_function=restore_simulation_state_with_activation_times

[assembly_matrix]
init_function=set_initial_conditions_fvm
sigma_l=0.0003
sigma_t=0.0001
sigma_n=0.00005
library_file=shared_libs/libdefault_matrix_assembly.so
main_function=anisotropic_sigma_assembly_matrix

[linear_system_solver]
tolerance=1e-15
use_preconditioner=no
max_iterations=100
library_file=shared_libs/libdefault_linear_system_solver.so
use_gpu=yes
main_function=gpu_biconjugate_gradient
init_function=init_gpu_biconjugate_gradient

[domain]
name=Cuboid Mesh
;micro meters
start_dx=200.0
start_dy=200.0
start_dz=200.0
side_length_x=1000
side_length_y=1000
side_length_z=1000
main_function=initialize_grid_with_cuboid_mesh

[ode_solver]
adaptive=true
dt=0.0001
use_gpu=true
gpu_id=0
library_file= shared_libs/libToRORd_fkatp_endo.so

[stim_plain]
start = 0.0
duration = 2.0
period=312.0
current = -80.0
x_limit = 200.0
main_function=stim_if_x_less_than

**Reply by LLRiebel** [reply]
Hi Rafael,

Thank you for the quick fix! 

I've done a test run and everything seems to be working well! 

Being able to compute the APD after restoring state is quite crucial but I can time the save point so it falls in between beats and the cells are definitely at relaxation at that point and the two new functions work :)  

Thanks again,
Luana

**Comment by adasi11** [comment]
Hi, 

For modeling scar tissue, I have seen that some changes are applied in matrix_assembly, such as:
ac->sigma.x = b_sigma_x*sigma_factor;   being sigma_factor close to zero

Are further changes needed in the cellular model? Such as setting all ionic parameters to zero?

I am trying to reproduce scar tissue and in some cells with very low sigma, the Vm progressively goes below -600 mV, crashing the simulation. 
This is the case even if I set in scar cells:
rDY[i] = 0;   for i = 0,..., NEQ-1 
So Vm would be theoretically constant over the entire simulation and yet, it reaches very low values. 

What is the best approach to model non-conductive cells?

Thanks,
Albert

**Reply by rsachetto** [reply]
Hi Albert, 

I think that this particular function that uses the sigma_factor was implemented by Lucas, so I do not know the ideas behind it.  
In the scar simulations that I do, I usually inactivate the cell setting the property active to false. This makes the cell totally non-conductive. 

If you want to only reduce the cell conductivity, without inactivating it, you could apply the sigma factor idea. But if you do that, the PDE matrix becomes non-symmetric and I think that is the reason why the simulation is crashing (you are probably using the Conjugate Gradient to solve the PDE).  To avoid this you will have to use another linear system solver like the gpu_biconjugate_gradient (do not forget to change the init and end functions to init_gpu_biconjugate_gradient and end_gpu_biconjugate_gradient.) 

My best,
Rafael

**Reply by adasi11** [reply]
Thanks for the prompt reply!  I will try both methods.
Best,
Albert

**Reply by LLRiebel** [reply]
Hi Rafael,

just been chatting to Albert about this. I've been using a sigma factor, which isn't drastically low, with the conjugate method and it's been working fine. Now I am wondering if I should be using the biconjugate gradient method instead. Would the non-symmetry in the PDE matrix cause artefacts or be numerically inaccurate in the conjugate gradient method? And would you recommend using one over the other for some scenarios?

Thanks!
Luana

**Reply by rsachetto** [reply]
Hi Luana, 

If you use the same sigma factor for all cells, the matrix would still be symmetric. But if you have different sigma factors, and use them in some particular cells, your matrix will not be symmetric anymore. The CG method is **guaranteed** to work when the matrix is symmetric/Hermitian and positive definite. However, it can work in some non-symmetric matrices. 

If the CG is converging, the final solution is correct. However, to be safe, I think you should use the bi conjugate gradient method.

**Reply by LLRiebel** [reply]
Great, thank you for the explanation! 

I've tried the method but there doesn't seem to be an end_gpu_biconjugate_gradient method in gpu_solvers_cublas_10.c and the one in gpu_solvers_cublas_11.c isn't detected ("end_gpu_biconjugate_gradient function not found in the provided in library shared_libs/libdefault_linear_system_solver.so").

**Reply by rsachetto** [reply]
Hum, it seems the code is missing from the gpu_solvers_cublas_10.c. The other one is not being detected because you are using cublas version 10. I will updated the code as soon as I can.

**Reply by LLRiebel** [reply]
Oh yes of course! Thanks!

**Reply by rsachetto** [reply]
For now, you can just omit the end_function, as it is not mandatory.

**Reply by adasi11** [reply]
Following on the issue of modelling scar tissue, only inactivating the cell (setting active to false) worked for me.

However, is there a way to inactivating a cell and still being able to display it on ParaView? It looks very odd not having a group of cells in the mesh. It seems as though the blood is going to leak into the pericardium 😅🤣

**Reply by rsachetto** [reply]
Hi Albert, how are you modeling the scar tissue? Are you using the FIBROTIC macro? If yes, I can change the save function to save the FIBROTIC cells.

**Reply by adasi11** [reply]
I wasn't using that macro, but I can adapt the code to use it now.

Thanks!
Albert

**Reply by rsachetto** [reply]
So you do not have any additional info in your mesh? We need the info in order to know if a cell was inactivated because it does not belong to the domain you if it belongs to the scar.

**Reply by adasi11** [reply]
My mesh has a column for tissue type (and now one for the fibrotic boolean).
Before I was using the macro TISSUE_TYPE to inactivate cells, setting a tissue type of 0 in scar cells. But I have added now an extra column to use the boolean FIBROTIC.

**Reply by rsachetto** [reply]
Hi Albert, 

For now, I changed the code and it is possible to save the scar cells only if they have the FIBROTIC info. I will think about a better way to achieve this functionality. Maybe the user could provide a function to inform the system which cells should be saved.

To enable it, you will have to update you code and add the **save_scar_cells=true** in the **save_result section**. 

Could you please test and tell me if everything is working?

**Reply by adasi11** [reply]
All working on my end. Looks beautiful now. 
Thanks Rafael!

**Comment by LLRiebel** [comment]
@rsachetto Apologies for bothering you again Rafael, I've just pulled the latest changes and simulations I ran yesterday are now not propagating anymore. I've ran the cuboid_ohara.ini example and it's broken too. Could you check if that is the same on your end and maybe one of the last commits created an issue?

**Reply by rsachetto** [reply]
This is strange. The last commits only changed the visualization tool and added a save mesh function for the purkinje system

**Reply by LLRiebel** [reply]
I'm so sorry, it turned out to be the source of all evil: merge conflict. Surprisingly, GitHub acted like like the automatic merge went through... It is all working again now, really sorry for troubling you, I'll make sure to double check next time.

**Reply by rsachetto** [reply]
No problem at all. Glad to know that everything is working :)

**Comment by rsachetto** [comment]
@LLRiebel Could you try to do:
```sh 
./build.sh clean
./build.sh
```

The cuboid_ohara.ini seems to be working for me here.
![screenrec001](https://user-images.githubusercontent.com/506938/137341130-2e5be41c-e2e5-470e-8407-c41a6ba71ee8.gif)

What is the result of the cuboid_ohara.ini for you?
